### PR TITLE
feat: Implement KeyPackage pruning [CL-16] Redux

### PR DESCRIPTION
**Pull request by OtaK**
* Previous PR here: #32

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Keypackages that were past their configured lifetime would stay in the keystore, leading to duplication and maybe multi-use of Keypackages

### Solutions

* Added an API to allow consumers to manually prune uploaded Keypackages from the Keystore
	* Added warnings that this can be misused and direct calls to the method implies that the consumer knows what they're doing
* Add an API to get the count of valid KeyPackages.
* Added automatic calls to the keypackage pruning method to ensure lifetime-expired keypackages are removed from the store
* ~~Added automatic keypackage pruning of keypackage refs covered by a Welcome message~~
	* Edit: This has been discarded as OpenMLS does it internally already
* WASM API fixes + Added rustdoc comments in some places
	* Moved all `Box<[u8]>` getters to concrete JS `Uint8Array` because sometimes `wasm-bindgen` wouldn't get it right and it would end up as a JS `Array` of `number` (between 0 and 255) making it extremely inefficient and confusing API-wise.
	* Fixed an issue with async double borrows on WASM causing panics when certain mutable operations occur simultaneously
* Fixed some edge case bugs here and there

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

```
cargo test client_automatically_prunes_lifetime_expired_keypackages
cargo test client_generates_correct_number_of_kpbs
cargo test conversation_from_welcome_prunes_local_keypackage
```

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by OtaK** [comment]
To do: Add an API to get the count of valid KeyPackages. CoreLogic will need it to inform the decision to upload new packages to the DS or not

File: crypto/src/client.rs [hunk 1]
```diff
@@ -264,6 +295,83 @@ impl Client {
...
+                valid_count += 1;
+            }
+
+            CryptoResult::Ok(valid_count)
+        })?;
+
+        Ok(valid_count)
+    }
+
+    fn is_mls_keypackage_expired(kpb: &KeyPackageBundle) -> bool {
```
Comment (beltram) [inline crypto/src/client.rs:320]: This should be unitary tested

File: crypto/src/lib.rs [hunk 2]
```diff
@@ -267,6 +267,10 @@ impl MlsCentral {
             .await
     }
 
+    pub async fn client_valid_keypackages_count(&self) -> CryptoResult<usize> {
```
Comment (beltram) [inline crypto/src/lib.rs:270]: documentation

File: crypto/src/client.rs [hunk 3]
```diff
@@ -264,6 +295,83 @@ impl Client {
...
+
+        let mut conn = keystore.borrow_conn().await?;
+
+        let kps = MlsKeypackage::find_all(&mut *conn, EntityFindParams::default()).await?;
+
+        let ids_to_delete = kps.into_iter().try_fold(Vec::new(), |mut acc, kp| {
+            let kpb = KeyPackageBundle::from_key_store_value(&kp.key).map_err(MlsError::from)?;
+            let mut is_expired = Self::is_mls_keypackage_expired(&kpb);
+            if !is_expired && !refs.is_empty() {
+                let href = kpb.key_package().hash_ref(crypto).map_err(MlsError::from)?;
```
[the remark this answers is not in the document]
Comment (OtaK) [reply crypto/src/client.rs:356]: Good catch, this will speed the method up dramatically

**Review by beltram — APPROVED** [review]
